test/gtest/ucp/test_ucp_tag.cc

```diff
@@ -423,22 +424,131 @@ class test_ucp_tag_limits : public test_ucp_tag {
+                          m_min_rndv);
+                EXPECT_EQ(cfg->tag.offload.max_eager_short.memtype_off + 1,
+                          m_min_rndv);
+            } else {
+                check_short_thresh(cfg->tag.offload.max_eager_short, cfg_thresh);
+            }
+        } else {
+            check_short_thresh(cfg->tag.max_eager_short, cfg_thresh);
+        }
+    }
+
+    void check_am_rndv_v2(size_t cfg_thresh)
+    {
+        ucp_ep_config_t *cfg = ucp_ep_config(sender().ep());
+
+        check_short_thresh(cfg->am_u.max_eager_short, cfg_thresh);
+        check_short_thresh(cfg->am_u.max_reply_eager_short, cfg_thresh);
+    }
+
+    void check_ep_proto_rndv(size_t cfg_thresh, bool expect_rndv)
+    {
+        ucp_ep_config_t *cfg             = ucp_ep_config(sender().ep());
+        ucp_proto_select_t *proto_select = &cfg->proto_select;
+        size_t msg_length                = cfg_thresh;
+        ucp_proto_select_elem_t value;
+
+        if (m_test_offload) {
+            /* There is a lower bound for rndv threshold with tag offload.
+             * We should not send messages smaller than this limit */
+            msg_length = ucs_max(m_min_rndv, msg_length);
+        }
+
+        kh_foreach_value(proto_select->hash, value, {
+            /* Find index of the corresponding ucp_proto_threshold_elem_t
+             * to handle the given message size */
+            unsigned idx = 0;
+            const ucp_proto_config_t *proto_config;
+            for (; msg_length > value.thresholds[idx].max_msg_length; ++idx) {
+                proto_config = &value.thresholds[idx].proto_config;
+                /* Assert no rndv before expected limit */
+                EXPECT_EQ(nullptr, strstr(proto_config->proto->name, "rndv"));
+            }
+            proto_config = &value.thresholds[idx].proto_config;
+
+            if (expect_rndv) {
+                EXPECT_EQ(proto_config->cfg_thresh, cfg_thresh);
+                EXPECT_NE(nullptr, strstr(proto_config->proto->name, "rndv"));
+            } else if (!m_test_offload) {
+                /* Skip check that rndv is disabled for tag offload use case.
+                 * With tag offload rndv protocol might still be used for large
+                 * messages, because of HWTM limitation for eager transfers. */
+                EXPECT_NE(proto_config->cfg_thresh, cfg_thresh);
+                EXPECT_EQ(nullptr, strstr(proto_config->proto->name, "rndv"));
+            }
+        });
+    }
+
+    void check_rndv_threshold(size_t cfg_thresh)
+    {
+        ucp_context_config_t *cfg = &sender().worker()->context->config.ext;
+        if (cfg->proto_enable) {
+            /* Check proto_v2 rndv thresholds only when this protocol is
+             * enabled, otherwise these checks are irrelevant */
+            check_tag_rndv_v2(cfg_thresh);
+            check_am_rndv_v2(cfg_thresh);
+        }
+        check_ep_proto_rndv(cfg_thresh, true);
```

Comment:
Wouldn't `check_ep_proto_rndv` fail in case if `cfg->proto_enable` is not true? As I undestand it fully relies on `proto_select` which exists only for `proto_enable`.

Reply:
Thanks for spotting! Well, these tests pass CI jobs with proto_disabled

But I think they pass because inside `check_ep_proto_rndv` we iterate over protocols hash map
```
        kh_foreach_value(proto_select->hash, value, {
```
which is supposedly empty in case of proto_disable. So essentially this test may test nothing for v1)

I will run this test with PROTO disabled and let you know the outcome.
If it's like that, then all runtime tests are v2 specific and I'll update the code accordingly